### PR TITLE
DISCUSSION: group_keys_by_replica behavior

### PR DESCRIPTION
**Pull request by mambocab**
I've done a little cleanup of some unimportant docs and formatting nits in this commit, but the important change is in the way it chooses a host for each key.

This logic:
```
for replica in valid_replicas:
    if replica in keys_per_host:
        keys_per_host[replica].append(key)
        break
else:		
        keys_per_host[random.choice(valid_replicas)].append(key)
```

will end up excluding some keys under certain conditions. I used [this script](https://github.com/mambocab/python-driver/files/1717574/test_gbr.py.txt) to demonstrate. I ran 

```sh
ccm create test-endpoints -n 5 -v 3.11.1 && ccm start --wait-for-binary-proto g co master ; python test_gbr.py | tee replica-output-master.txt ; g co proposed-grouping-changes; python test_gbr.py | tee replica-output-proposed.txt
```

What happens in this test (5 nodes, RF=3) is this: `group_keys_by_replica` checks C* keys and adds them, preferring hosts that are already there, until there are 3 hosts in `keys_per_host`. Once those 3 hosts are chosen, no new hosts will be added -- with RF=3 across 5 nodes, every key is guaranteed to map to `>= 1` of the 3 hosts already in the map. See the attached output: ([replica-output-master.txt](https://github.com/mambocab/python-driver/files/1717577/replica-output-master.txt), [replica-output-proposed.txt](https://github.com/mambocab/python-driver/files/1717578/replica-output-proposed.txt))

This is technically working as documented, but I don't think it's the intended behavior -- in cases like these, it doesn't end up distributing keys to a subset of hosts, and it doesn't distribute to that subset very evenly (since, in this case, one host gets a disproportionately large section of the token ring).

The proposal in this PR, which I haven't given too much thought, is always-random assignment. Was there a reason we didn't go with this initially?

### CODE REVIEW

File: cassandra/metadata.py [hunk 1]
```diff
@@ -2576,46 +2576,45 @@ def _cql_from_cass_type(cass_type):
...
+    Returns a :class:`dict` with the keys grouped per host. This can be
     used to more accurately group by IN clause or to batch the keys per host.
 
     If a valid replica is not found for a particular key it will be grouped under
     :class:`~.NO_VALID_REPLICA`
 
     Example usage::
-        result = group_keys_by_host(session, "system", "peers",
-                ("peer", "data_center"), (("127.0.0.1", ), ("127.0.0.2", )))
+        result = group_keys_by_host(
```
Comment (beltran) [inline cassandra/metadata.py:2586]: Good catch in the docs

File: cassandra/metadata.py [hunk 2]
```diff
@@ -2576,46 +2576,45 @@ def _cql_from_cass_type(cass_type):
...
         if len(serialized_key) == 1:
             routing_key = serialized_key[0]
         else:
             routing_key = b"".join(struct.pack(">H%dsB" % len(p), len(p), p, 0) for p in serialized_key)
         all_replicas = cluster.metadata.get_replicas(keyspace, routing_key)
         # First check if there are local replicas
         valid_replicas = [host for host in all_replicas if
-                              host.is_up and cluster._default_load_balancing_policy.distance(
-                                  host) == HostDistance.LOCAL]
+                          host.is_up and distance(host) == HostDistance.LOCAL]
```
Comment (beltran) [inline cassandra/metadata.py:2609]: :+1:

**Comment by beltran** [comment]
Good catch! I think it makes more sense like this.
The reason it was made like that originally was to maximize the grouping. And the reason for that was to avoid inter node communication if a node was queried with that set of keys but this is avoided as well this way and the load is more evenly distributed.

**Comment by mambocab** [comment]
Cool, thanks for the reviews. Will clean up the history and submit a PR!

**Pull request closed without merging**